src/test/kotlin/org/rust/lang/core/type/RsIndexTypeInferenceTest.kt

```diff
@@ -129,4 +129,61 @@ class RsIndexTypeInferenceTest : RsTypificationTestBase() {
           //^ i32
         }
     """)
+
+    fun testDeref() = testExpr("""
+        #[lang = "index"]
+        pub trait Index<Idx: ?Sized> {
+            type Output: ?Sized;
+            fn index(&self, index: Idx) -> &Self::Output;
+        }
+
+        struct Vec<T>;
+
+        impl<T> Index<usize> for Vec<T> {
+            type Output = T;
+            fn index(&self, index: usize) -> &T { unimplemented!() }
+        }
+
+        fn foo(v: Vec<u32>) -> u32 {
+            let p = &v;
+            let m = p[0];
+            m
+          //^ u32
+        }
+    """)
+
+    fun testTransitiveCoercion() = testExpr("""
+        #[lang = "deref"]
+        pub trait Deref {
+            type Target: ?Sized;
+            fn deref(&self) -> &Self::Target;
+        }
+
+        #[lang = "index"]
+        pub trait Index<Idx: ?Sized> {
+            type Output: ?Sized;
+            fn index(&self, index: Idx) -> &Self::Output;
+        }
```

Comment:
Let's do the same trick here as for `org.rust.lang.core.resolve.RsClosuresResolveTest` to avoid repeating index impl every time?